update state-res to handle room versions 7, 8, 9

DevinR528 opened the pull request:
Fixes https://github.com/ruma/ruma/issues/718

Still kinda WIP and should probably add tests for the features specifically.

DevinR528 committed:
state-res: Add membership check for Knock behind unstable feature

DevinR528 commented on crates/ruma-state-res/src/room_version.rs:
```diff
@@ -157,6 +177,61 @@ impl RoomVersion {
+            extra_redaction_checks: true,
+            // Wait so should this be false?
+            // because: https://github.com/matrix-org/matrix-doc/pull/3289#issuecomment-884165177
+            allow_knocking: false,
```
See the comment in the link https://github.com/matrix-org/matrix-doc/pull/3289#issuecomment-884165177

jplatte replied:
To me it seems like the comment says the opposite, that this should be `true` so v8 is a simple extension of v7 rather than both being different extensions of v6.

DevinR528 commented on crates/ruma-state-res/src/event_auth.rs:
```diff
@@ -557,6 +559,35 @@ fn valid_membership_change(
         }
+        // TODO: Does this need to be gated
+        #[cfg(feature = "unstable-pre-spec")]
+        MembershipState::Knock if room_version.allow_knocking => {
```
Should this be gated, `MembershipState::Knock` is not.

jplatte replied:
Well, `room_version.allow_knocking` is feature-gated so this has to be. Otherwise it won't compile.

DevinR528 replied:
Good point. Should the variant also be gated as it is only part of the unstable spec or is that close enough?

jplatte replied:
The variant has been part of the stable spec for a long time, just without any meaning attached.

DevinR528 commented on crates/ruma-events/src/room/member.rs:
```diff
@@ -93,6 +93,13 @@ pub struct RoomMemberEventContent {
+    #[cfg_attr(docsrs, doc(cfg(feature = "unstable-pre-spec")))]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[ruma_event(skip_redaction)]
+    pub join_authorised_via_users_server: Option<UserId>,
```
Ok so `#[ruma_event(skip_redaction)]` is actually wrong since: in room v8 this wasn't true but 9 fixes this. Oh wait I can just do what the aliases event does right.

jplatte replied:
Yeah, this requires a custom redaction impl like with aliases.

jplatte commented on crates/ruma-state-res/src/room_version.rs:
```diff
@@ -61,6 +61,10 @@ pub struct RoomVersion {
     pub extra_redaction_checks: bool,
+    /// Allow knocking in event authentication.
+    pub allow_knocking: bool,
+    /// Allow knocking in event authentication.
```
Copy-paste error.

jplatte commented on crates/ruma-state-res/src/event_auth.rs:
```diff
@@ -11,6 +11,7 @@ use ruma_events::{
     },
     EventType,
 };
+
```
There shouldn't be an extra newline here.

jplatte commented on crates/ruma-state-res/src/event_auth.rs:
```diff
@@ -557,6 +617,35 @@ fn valid_membership_change(
+        MembershipState::Knock if room_version.allow_knocking => {
+            // 1. If the `join_rule` is anything other than `knock`, reject.
+            if join_rules != JoinRule::Knock {
+                warn!("Join rule is not set to knock");
```
This warning message is rather generic.

jplatte commented on crates/ruma-state-res/src/event_auth.rs:
```diff
@@ -557,6 +617,35 @@ fn valid_membership_change(
+                // allow.
+                // 4. Otherwise, reject.
+                if sender != target_user {
+                    warn!("Can't make other user join");
```
This one even more so.

jplatte commented on crates/ruma-state-res/src/room_version.rs:
```diff
@@ -72,6 +76,12 @@ impl RoomVersion {
             RoomVersionId::Version5 => Self::version_5(),
             RoomVersionId::Version6 => Self::version_6(),
+            #[cfg(feature = "unstable-pre-spec")]
+            RoomVersionId::Version7 => Self::version_7(),
```
For this to work reliably, state-res' `unstable-pre-spec` feature has to activate identifiers' pre-spec feature.

jplatte commented on crates/ruma-state-res/src/room_version.rs:
```diff
@@ -157,6 +177,61 @@ impl RoomVersion {
+            allow_knocking: true,
+            restricted_join_rules: false,
+        }
+    }
```
Can we change all of these to be associated constants instead, and also use the form `Self { version: VersionX, new_setting: true, ..Self::VERSION_BEFORE }`?

DevinR528 replied:
Such a good idea!! I feel like I tried consts at some point but don't remember why I didn't use them.

jplatte commented on crates/ruma-state-res/src/event_auth.rs:
```diff
@@ -455,7 +503,20 @@ fn valid_membership_change(
+                        //
+                        // 2. The join event has a valid signature from a homeserver whose
+                        // users have the power to issue invites.
+                        && restricted_join_rules_auth();
```
The indentation is kind of misleading here, I think `allow` should be split into named (as in local variables) sub-conditions. Also I don't get why there's a closure being passed here.

DevinR528 replied:
Ugh, I thought I commented on this a long time ago.

It's a closure to avoid having to pass the `fetch_state` closure, `content`, `auth_user` and maybe one other ID or event. I know this isn't ideal but I was unsure of how else to add all the complexity for a "cold" path that would become hot eventually.

jplatte replied:
I think it makes more sense to just pass these things. Maybe some further refactoring can be done later.

DevinR528 replied:
Should I also not worry about the amount of upfront work I end up doing (calling `fetch_state` a few times) even though I may not use any of them if the event isn't the type I have to check (if it's not a member event of a room >= 8?
EDIT: nevermind it really isn't that bad, it was only one possibly extra deserialization and no exrta fetch_state calls.

jplatte replied:
If you need intermediate bindings, I'd rather introduce a little bit of nesting than doing unnecessary upfront work.

DevinR528 commented:
Same here, I will try and get to it in the next few days.

jplatte commented on crates/ruma-events/src/room/member.rs:
```diff
@@ -107,10 +116,89 @@ impl RoomMemberEventContent {
+    /// Create a `RedactedRoomMemberEventContent` with the given membership and auth_user.
+    ///
+    /// This is only valid for room version 9 and above.
+    pub fn new_v9(membership: MembershipState, auth_user: UserId) -> Self {
```
Missing cfg attribute for this function.

DevinR528 committed:
state-res: Clean up the restrict join rules checks (remove closure)

DevinR528 commented:
The 2 tests I added are mostly just a sanity check. It really needs to be tested in conduit (although Nyaaori (matrix user in the conduit room) tested the old code and said that worked). It's too bad it's SO hard to test state-res, in general, it would be nice to have some peace of mind when you make a change :shrug:.

DevinR528 committed:
state-res: Make room version associated consts

DevinR528 committed:
state-res: Add support for room version 8 join restrictions

DevinR528 committed:
state-res: Fix logic of joining restricted rooms

We were checking the power level of the target user, we need to check
that the user selected to authorise the join in a restricted room when
the room version is >= 8.

jplatte commented:
I think this was only discussed on Matrix, so for clarity: I'm currently waiting on Devin to split the events change into a separate PR.

DevinR528 commented:
Hmm I thought someone was going to cherry-pick :laughing: ;)

I'm just running the tests locally and I'll push a new PR now.

jplatte commented:
Merged the knocking part and did some other refactorings. There's an updated version of the changes here: https://github.com/ruma/ruma/tree/add-roomver

DevinR528 commented:
Should I continue from the linked branch and open a new PR or doesn't matter? Thanks for working on this piecemeal with me, and letting it just sit for a while.

jplatte commented:
You can reset your branch to that one, start a new PR from it directly or whatever works for you.

jplatte commented:
Cleaned up and merged the remaining changes:

* 7290860019258f305da2c6b509c208196e442fa5
* 2f732bf9c2a83c960d4d8c9469fd42b10d53b361

aaronraimist commented:
@jplatte so is https://github.com/ruma/ruma/issues/718 done?

jplatte commented:
Yeah, I've closed that issue.
